File: 1652.defuse-the-bomb.py
#
# @lc app=leetcode id=1652 lang=python3
#
# [1652] Defuse the Bomb
#

# @lc code=start
class Solution:
    def decrypt(self, code: list[int], k: int) -> list[int]:

        #! I will use the brutal force first to check if it is right
        #! The time complexity is O(n + k)
        # result = [0]*len(code)
        # sum = 0

        # if k == 0:
        #     return result
        # if k > 0:
        #     #now we have the proper list
        #     for i in range(k):
        #         code.append(code[i])
            
        #     for i in range(len(code)):
        #         if i == 0:
        #             continue
        #         sum += code[i] 
        #         if i < k:
        #             continue
        #         result[i-k] = sum
        #         sum -= code[i-k+1]
        #     return result
        # else:
        #     for i in range(abs(k)):
        #         code.insert(0, code[-1-i])
        #     for i in range(len(code)- 1):
        #         sum += code[i]
        #         if i < abs(k) - 1:
        #             continue
        #         result[i-abs(k)+ 1] = sum
        #         sum -= code[i-abs(k)+ 1]
        #     return result
        #! The brutal force is correct, but it is not efficient
        #! What I miss is the trick to handle the circualr array
        #! We can use the modulo to fix the circular array
        #! Together with the slicing window, we can solve the problem

        result = [0]*len(code)
        if k == 0: 
            return result
        begin, end, window_sum = 1, k, 0
        if k < 0:
            begin, end = len(code) + k, len(code) - 1
        #Calculate the initial window sum
        for i in range (begin, end+1):
            window_sum += code[i]
        for i in range(len(code)):
            result[i] = window_sum
            #Here we use the module to hanlde the circular array
            #And thus updating the window sum
            window_sum -= code[begin % len(code)]
            window_sum += code[(end+1) % len(code)]
            begin += 1
            end += 1 
        return result       
# @lc code=end

File: 1652.defuse-the-bomb.java
/*
 * @lc app=leetcode id=1652 lang=java
 *
 * [1652] Defuse the Bomb
 */
class Solution {
    public int[] decrypt(int[] code, int k) {
        // Brutal force solution
        // List<Integer> code_integer = new ArrayList<>();
        // int[] result = new int[code.length];
        // for(int i = 0; i < code.length; i++){
        //     code_integer.add(code[i]);
        // }
        // if(k == 0){
        //     return result;
        // }
        
        // if (k > 0){
        //     for(int i = 0; i < k; i++){
        //         code_integer.add(code_integer.get(i));
        //     }
        //     for(int i = 0; i < code.length; i++){
        //         int sum = 0;
        //         for(int j = i + 1; j <= i + k; j++){
        //             sum += code_integer.get(j);
        //         }
        //         result[i] = sum;
        //     }
        // }else{
        //     for(int i = 0; i < -k; i++){
        //         code_integer.add(0, code_integer.get(code_integer.size() -1 - i));
        //     }
        //     for(int i = 0; i < code.length; i++){
        //         int sum = 0;
        //         for(int j = i; j < i-k; j++){
        //             sum += code_integer.get(j);
        //         }
        //         result[i] = sum;
        //     }
        // }
        // return result;

        // Better solution
        // int n = code.length;
        // int[] result = new int[n];
        // if (k == 0) {
        //     return result;
        // }
        // int right_start = k > 0 ? k + 1 : n; // confirm the right start index of the sliding window
        // k = Math.abs(k);
        // int sum = 0;
        // for(int i = right_start - k; i < right_start; i++){
        //     sum += code[i];
        // }
        // for(int i = 0; i < n; i++){
        //     result[i] = sum;
        //     sum += code[right_start % n] - code[(right_start - k) % n];
        //     right_start++;
        // }
        // return result;

        //! The answers are formed by continuous subarrays of length k -> Enumeration and sliding window technique.
        //TODO the sliding window is used to avoid repeated calculation of sum and validdate the status of the window
        //? the modular arithmetic is used to simulate the circular array
        int n = code.length;
        int[] ans = new int[n];
        if(k == 0) return ans;
        int rightEnd = 0;
        if(k > 0){
            rightEnd = k + 1;
        }
        else{
            rightEnd = n;
            k = -k;
        }

        int sum = 0;
        for(int i = rightEnd - k; i < rightEnd; i++){
            sum += code[i];
        }
        for(int i = 0; i < n; i++){
            ans[i] = sum;
            sum += code[rightEnd % n];
            sum -= code[(rightEnd - k) % n];
            rightEnd++;
        }
        return ans;
        



    }
}
// @lc code=end

